experiment/molecule/src/common.ts

```diff
@@ -14,55 +15,303 @@ export const Uint8: FixedBinaryCodec<number> = {
+export const HexUint128LE: FixedBinaryCodec<string> = createBIHexCodec(
+  Uint128LE
+);
+export const HexUint128BE: FixedBinaryCodec<string> = createBIHexCodec(
+  Uint128BE
+);
+/**
+ * @alias HexUint128LE
+ */
+export const HexUint128 = HexUint128LE;
 
 // array Uint256 [byte; 32]
-export declare const Uint256LE: FixedBinaryCodec<BI>;
-export declare const HexUint256LE: FixedBinaryCodec<string>;
+export const Uint256LE: FixedBinaryCodec<BI> = createBICodec(32);
+export const Uint256BE: FixedBinaryCodec<BI> = createBICodec(32, true);
+/**
+ * @alias Uint256LE
+ */
+export const Uint256 = Uint256LE;
+export const HexUint256LE: FixedBinaryCodec<string> = createBIHexCodec(
+  Uint256LE
+);
+export const HexUint256BE: FixedBinaryCodec<string> = createBIHexCodec(
+  Uint256BE
+);
+/**
+ * @alias HexUint256LE
+ */
+export const HexUint256 = HexUint256LE;
 
 // array Uint512 [byte; 64]
-export declare const Uint512LE: FixedBinaryCodec<BI>;
-export declare const HexUint512LE: FixedBinaryCodec<string>;
+export const Uint512LE: FixedBinaryCodec<BI> = createBICodec(64);
+export const Uint512BE: FixedBinaryCodec<BI> = createBICodec(64, true);
+/**
+ * @alias Uint512LE
+ */
+export const Uint512 = Uint512LE;
+export const HexUint512LE: FixedBinaryCodec<string> = createBIHexCodec(
+  Uint512LE
+);
+export const HexUint512BE: FixedBinaryCodec<string> = createBIHexCodec(
+  Uint512BE
+);
+/**
+ * @alias HexUint512LE
+ */
+export const HexUint512 = HexUint512LE;
 
 // array BytesN [byte; n]
-export declare function fixedHexBytes(
-  byteLength: number
-): FixedBinaryCodec<string>;
+export function fixedHexBytes(byteLength: number): FixedBinaryCodec<string> {
+  return {
+    __isFixedCodec__: true,
+    byteLength,
+    pack(hexStr: string) {
+      assertHexString(hexStr);
+      const result = toArrayBuffer(hexStr);
+      if (byteLength > 0 && byteLength !== result.byteLength) {
+        throw new Error(
+          `Invalid hex string length: ${result.byteLength}, should be ${byteLength}`
+        );
+      }
+      return result;
+    },
+    unpack(buf) {
+      return serializeJson(buf);
+    },
+  };
+}
 
 // vector Bytes <byte>
-export declare const HexBytes: BinaryCodec<string>;
-export declare const UTF8String: BinaryCodec<string>;
+export const HexBytes: BinaryCodec<string> = {
+  pack(hexStr: string) {
+    assertHexString(hexStr);
+    return toArrayBuffer(hexStr);
+  },
+  unpack(buf) {
+    return serializeJson(buf);
+  },
+};
+export const UTF8String: BinaryCodec<string> = {
+  pack(hexStr: string) {
+    assertHexString(hexStr);
+    const result = toArrayBuffer(hexStr);
+    return result;
+  },
+  unpack(buf) {
+    return serializeJson(buf);
+  },
+};
+
+export function createBICodec(
+  byteLength: number,
+  bigEndian?: boolean
+): FixedBinaryCodec<BI> {
+  const loops = Math.ceil(byteLength / 4);
+  return {
+    __isFixedCodec__: true,
+    byteLength,
+    pack(num: BI) {
+      const buffer = new ArrayBuffer(byteLength);
```

Comment:
We should first check that `num` is legal and that `byteLength` is sufficient for storage the `num`

Reply:
done